File: c/problem_58.c
#include "problem_58.h"
#include <stdio.h>
#include <assert.h>

int search_rotated(long long *arr, int n, long long target) {
    int lo=0, hi=n-1;
    while (lo<=hi) {
        int mid=(lo+hi)/2;
        if (arr[mid]==target) return mid;
        if (arr[lo]<=arr[mid]) {
            if (arr[lo]<=target && target<arr[mid]) hi=mid-1; else lo=mid+1;
        } else {
            if (arr[mid]<target && target<=arr[hi]) lo=mid+1; else hi=mid-1;
        }
    }
    return -1;
}

void problem58(void) {
    printf("Problem 58:\n");
    long long a[] = {13,18,25,2,8,10};
    assert(search_rotated(a,6,8)==4);
    assert(search_rotated(a,6,13)==0);
    assert(search_rotated(a,6,99)==-1);
    printf("All Tests Passed\n");
}

File: lua/problem_58.lua
local M = {}

function M.search_rotated(arr, target)
    local lo, hi = 1, #arr
    while lo <= hi do
        local mid = math.floor((lo + hi) / 2)
        if arr[mid] == target then return mid end
        if arr[lo] <= arr[mid] then
            if arr[lo] <= target and target < arr[mid] then hi = mid - 1
            else lo = mid + 1 end
        else
            if arr[mid] < target and target <= arr[hi] then lo = mid + 1
            else hi = mid - 1 end
        end
    end
    return nil
end

function M.problem58()
    print("Problem 58:")
    assert(M.search_rotated({13,18,25,2,8,10}, 8)  == 5)
    assert(M.search_rotated({13,18,25,2,8,10}, 13) == 1)
    assert(M.search_rotated({13,18,25,2,8,10}, 99) == nil)
    print("All Tests Passed")
end

return M
